### PR TITLE
Fixes #34872 - Search for errata tasks by scoped search

### PR DESCRIPTION
**Pull request by adamruzicka**
Requires:
- https://github.com/theforeman/foreman_remote_execution/pull/719

#### What are the changes introduced in this pull request?
When looking up tasks to use as data sources when generating errata application report, Katello had to implement quite a lot of the logic to match the right tasks. Recently, REX has added support for searching on tasks using a REX feature, which can be reused here to drop the all the custom code.

#### Considerations taken when implementing this change?
Primary drive here was removal of custom code while making sure thing still work the same way

#### What are the testing steps for this pull request?
- Apply this patch and the patch in REX
- try generating an errata application report.

### CODE REVIEW

File: app/lib/katello/concerns/base_template_scope_extensions.rb [hunk 1]
```diff
@@ -206,18 +206,15 @@ def load_errata_applications(filter_errata_type: nil, include_last_reboot: 'yes'
...
+          new_labels = 'label = "Actions::RemoteExecution::RunHostJob" AND remote_execution_feature.label = katello_errata_install'
+          labels = [labels, new_labels].map { |label| "(#{label})" }.join(' OR ')
         end
 
+        search = [search_up_to, search_since, search_result, "state = 'stopped'", label_search].compact.join(' and ')
+
         tasks = load_resource(klass: ForemanTasks::Task,
-                              where: condition,
                               permission: 'view_foreman_tasks',
                               joins: 'LEFT OUTER JOIN template_invocations ON foreman_tasks_tasks.id = template_invocations.run_host_job_task_id LEFT OUTER JOIN templates ON template_invocations.template_id = templates.id',
```
Comment (adamruzicka) [inline app/lib/katello/concerns/base_template_scope_extensions.rb:219]: How does the join work when REX is not available? I'd expect the database would complain about `template_invocations` table not being there.

File: app/lib/katello/concerns/base_template_scope_extensions.rb [hunk 2]
```diff
@@ -206,21 +206,22 @@ def load_errata_applications(filter_errata_type: nil, include_last_reboot: 'yes'
...
         search_result = status.present? && status != 'all' ? "result = #{status}" : nil
-        search = [search_up_to, search_since, search_result].compact.join(' and ')
+        labels = 'label ^ (Actions::Katello::Host::Erratum::Install, Actions::Katello::Host::Erratum::ApplicableErrataInstall)'
+        select = 'foreman_tasks_tasks.*'
 
         if Katello.with_remote_execution?
-          condition = ["state = 'stopped' AND ((label = 'Actions::RemoteExecution::RunHostJob' AND templates.id = ?) " \
-                       "OR label = 'Actions::Katello::Host::Erratum::Install' OR label = 'Actions::Katello::Host::Erratum::ApplicableErrataInstall')",
-                       RemoteExecutionFeature.feature('katello_errata_install').job_template_id]
+          new_labels = 'label = Actions::RemoteExecution::RunHostJob AND remote_execution_feature.label = katello_errata_install'
```
Comment (jeremylenz) [inline app/lib/katello/concerns/base_template_scope_extensions.rb:213]: The remote execution feature can be either `katello_errata_install` or `katello_errata_install_by_search`. The former is used on the old pages, and `katello_errata_install_by_search` is used on the new host detail page.

I'm surprised no one caught this sooner..

**Review by jeremylenz — COMMENTED** [review]
I was trying to figure out why I was getting a blank report, even though I had applied some errata just a minute ago. Then I discovered why (see below)..

File: app/lib/katello/concerns/base_template_scope_extensions.rb (hunk 2 above)
Comment (adamruzicka) [reply app/lib/katello/concerns/base_template_scope_extensions.rb]: Good catch, added

**Comment by jeremylenz** [comment]
I've always thought it was kind of an insane hack that when we try to address the user need of "I want to know what errata were applied on my host" our solution is just to look for the completed tasks. It's always seemed fragile and sketchy. Clearly the "correct" way to do it is to store something every time we apply an errata successfully.

Storing it as a comment in the template is an innovative solution and quick and easy at the same time. I wouldn't have thought of it! I'm fine that it won't work for historical errata applications made with this template because this report has never worked for those same applications anyway.

In the long run I'd prefer to do it the right way and store it as a database model. We already do it with export histories, so I don't see why we couldn't do the same for errata applications. This way we could treat every errata application the same way regardless of job template, and stop relying on tasks once and for all. (cc @parthaa, would also be good to get your thoughts)

**Comment by adamruzicka** [comment]
> Storing it as a comment in the rendered template is an innovative solution and quick and easy at the same time

And also somewhat hacky :) 

> In the long run I'd prefer to do it the right way and store it as a database model.

I'd like to see that in general, not just in here. Treating tasks as disposable would be nice

**Comment by adamruzicka** [comment]
Hope the react test failures aren't related

**Comment by jeremylenz** [comment]
I just applied 94 errata to my host, but I'm still getting a blank report.

When I look at the job and go to Preview template, I'm not seeing the IDs stored there.

Is it maybe because it calls `render_template('Package Action - Script Default')` and that's the template I see rendered?

Here's what it looks like
```

    
    #!/bin/bash

# Helper function that exits with a particular message and code.
#
# Usage:
#   exit_with_message "Could not do a thing" 2
#
exit_with_message () {
  echo "${1}, exiting..."
  exit $2
}



# Action
  yum -y update-minimal --advisory=RHBA-2022:1984 --advisory=RHBA-2022:1992 --advisory=RHBA-2022:1993 --advisory=RHBA-2022:1995 --advisory=RHBA-2022:1996 --advisory=RHBA-2022:2000 --advisory=RHBA-2022:2002 --advisory=RHBA-2022:2004 --advisory=RHBA-2022:2005 --advisory=RHBA-2022:2006 --advisory=RHBA-2022:2007 --advisory=RHBA-2022:2015 --advisory=RHBA-2022:2016 --advisory=RHBA-2022:2021 --advisory=RHBA-2022:2025 --advisory=RHBA-2022:2027 --advisory=RHBA-2022:2034 --advisory=RHBA-2022:2038 --advisory=RHBA-2022:2039 --advisory=RHBA-2022:2040 --advisory=RHBA-2022:2042 --advisory=RHBA-2022:2044 --advisory=RHBA-2022:2045 --advisory=RHBA-2022:2046 --advisory=RHBA-2022:2047 --advisory=RHBA-2022:2048 --advisory=RHBA-2022:2049 --advisory=RHBA-2022:2050 --advisory=RHBA-2022:2051 --advisory=RHBA-2022:2052 --advisory=RHBA-2022:2053 --advisory=RHBA-2022:2054 --advisory=RHBA-2022:2055 --advisory=RHBA-2022:2060 --advisory=RHBA-2022:2061 --advisory=RHBA-2022:2063 --advisory=RHBA-2022:2064 --advisory=RHBA-2022:2065 --advisory=RHBA-2022:2067 --advisory=RHBA-2022:2068 --advisory=RHBA-2022:2069 --advisory=RHBA-2022:2070 --advisory=RHBA-2022:2071 --advisory=RHBA-2022:2072 --advisory=RHBA-2022:2075 --advisory=RHBA-2022:2076 --advisory=RHBA-2022:2077 --advisory=RHBA-2022:2078 --advisory=RHBA-2022:2079 --advisory=RHBA-2022:2080 --advisory=RHBA-2022:2082 --advisory=RHBA-2022:2083 --advisory=RHBA-2022:2085 --advisory=RHBA-2022:2086 --advisory=RHBA-2022:2090 --advisory=RHBA-2022:2095 --advisory=RHBA-2022:2096 --advisory=RHBA-2022:2097 --advisory=RHBA-2022:2099 --advisory=RHBA-2022:2100 --advisory=RHBA-2022:2101 --advisory=RHBA-2022:2102 --advisory=RHBA-2022:2103 --advisory=RHBA-2022:2104 --advisory=RHBA-2022:2105 --advisory=RHBA-2022:2106 --advisory=RHBA-2022:2107 --advisory=RHBA-2022:2112 --advisory=RHBA-2022:2113 --advisory=RHBA-2022:2114 --advisory=RHBA-2022:2117 --advisory=RHBA-2022:2118 --advisory=RHBA-2022:2119 --advisory=RHBA-2022:2121 --advisory=RHBA-2022:2122 --advisory=RHBA-2022:2147 --advisory=RHEA-2022:1985 --advisory=RHEA-2022:2014 --advisory=RHEA-2022:2029 --advisory=RHSA-2022:1546 --advisory=RHSA-2022:1552 --advisory=RHSA-2022:1642 --advisory=RHSA-2022:1891 --advisory=RHSA-2022:1932 --advisory=RHSA-2022:1986 --advisory=RHSA-2022:1988 --advisory=RHSA-2022:1991 --advisory=RHSA-2022:2008 --advisory=RHSA-2022:2013 --advisory=RHSA-2022:2031 --advisory=RHSA-2022:2043 --advisory=RHSA-2022:2110 --advisory=RHSA-2022:2201
RETVAL=$?
[ $RETVAL -eq 0 ] || exit_with_message "Package action failed" $RETVAL
```

**Comment by jeremylenz** [comment]
I re-did my seeds and I now see the applied errata in the template. :heavy_check_mark: 

But

when I run the report, I get an error
```
Field 'remote_execution_feature.label' not recognized for searching!
```

Also, if I don't add `&.` in
 https://github.com/theforeman/foreman_remote_execution/blob/291df12dee06d1a15852e3531191bbb869a300d6/app/helpers/remote_execution_helper.rb#L249-L250,

I get

Action-Controller-Exception-caught

**Comment by jeremylenz** [comment]
and when troubleshooting that error I was also confused by this:
```
[1] pry(main)> Setting['remote_execution_job_invocation_report_template']
=> "Job invocation - report template"
[2] pry(main)> ReportTemplate.all.map(&:name)
=> ["Ansible - Ansible Inventory",
 "Host - Applicable Errata",
 "Host - Applied Errata",
 "Host - compare content hosts packages",
 "Host - Enabled Repositories",
 "Host - Last Checkin",
 "Host - Registered Content Hosts",
 "Host - Statuses",
 "Job - Invocation Report",
 "Subscription - Entitlement Report",
 "Subscription - General Report",
 "User - Registered Users"]
[3] pry(main)> Setting['remote_execution_job_invocation_report_template'] = "Job - Invocation Report"
=> "Job - Invocation Report"
[4] pry(main)> ReportTemplate.where(name: Setting['remote_execution_job_invocation_report_template']).first
=> #<ReportTemplate:0x00007f324a1f64a8
 id: 145,
 name: "Job - Invocation Report",
 template:
  "<%#\nname: Job - Invocation Report\nsnippet: false\ntemplate_inputs:\n- name: job_id\n  required: true\n  input_type: user\n  description: Id of job invocation to report\n  advanced: false\n  value_type: plain\n  resource_type: JobInvocation\n  hidden_value: false\n- name: hosts\n  required: false\n  input_type: user\n  description: Field for filter hosts of job invocation. Leave blank for all hosts.\n  advanced: false\n  value_type: search\n  resource_type: Host\n  hidden_value: false\nmodel: ReportTemplate\nrequire:\n- plugin: foreman_remote_execution\n￼ version: 4.4.0\n-%>\n<%- report_headers 'Host', 'stdout', 'stderr', 'debug', 'Result', 'Finished' -%>\n<%- invocation = find_job_invocation_by_id(input('job_id')) -%>\n<%- parts = [\"job_invocation.id = \#{input('job_id')}\"] -%>\n<%- parts << input('hosts') unless input('hosts').blank? -%>\n<%- search = parts.map { |part| '(' + part + ')' }.join(' AND ') -%>\n<%- load_hosts(search: search).each do |batch| -%>\n<%-   batch.each do |host|  -%>\n<%-     task = invocation.sub_task_for_host(host) -%>\n<%-     outputs = { 'stdout' => [], 'stderr' => [], 'debug' => [] }  -%>\n<%-     task.action_continuous_output.each do |output| -%>\n<%-       outputs[output['output_type']] << output['output'] -%>\n<%-     end -%>\n<%-     report_row(\n            'Host': host.name,\n            'stdout': join_with_line_break(outputs['stdout']),\n            'stderr': join_with_line_break(outputs['stderr']),\n            'debug': join_with_line_break(outputs['debug']),\n            'Result': task.result,\n            'Finished': task.ended_at.blank? ? nil : format_time(task.ended_at),\n        ) -%>\n<%-   end -%>\n<%- end -%>\n<%= report_render -%>\n",
 snippet: false,
 template_kind_id: nil,
 created_at: Fri, 18 Mar 2022 18:11:04.988440000 UTC +00:00,
 updated_at: Fri, 27 May 2022 17:21:48.648646000 UTC +00:00,
 locked: true,
 default: true,
 vendor: "Foreman",
[5] pry(main)> Setting['remote_execution_job_invocation_report_template']
=> "Job - Invocation Report"
```

Changing the setting to match the report template didn't resolve it, for some reason. It was only adding the `&.` that helped.

**Comment by adamruzicka** [comment]
> when I run the report, I get an error
> 
> ```
> Field 'remote_execution_feature.label' not recognized for searching!
> ```

This requires some changes in REX, PR proposing those changes is linked in the description.

> Changing the setting to match the report template didn't resolve it, for some reason

Possibly a combination of things. The report template got renamed in foreman (https://github.com/theforeman/foreman/pull/9093), but a corresponding change in REX (https://github.com/theforeman/foreman_remote_execution/pull/692) is still pending and unless you have new enough foreman (containing https://github.com/theforeman/foreman/pull/9246), then changing settings manually doesn't really work.

**Review by jeremylenz — APPROVED** [review]
errata_report

Checked out the https://github.com/theforeman/foreman_remote_execution/pull/719 and was finally able to run the report successfully.

Let's get this in and move on! ACK :+1:

**Comment by jeremylenz** [comment]
Should I wait for https://github.com/theforeman/foreman_remote_execution/pull/719 to be in before merging?

**Comment by adamruzicka** [comment]
> Should I wait for https://github.com/theforeman/foreman_remote_execution/pull/719 to be in before merging?

Yes, that would be better. If you leave me a +1 there I can merge it straight away

**Comment by jeremylenz** [comment]
> If you leave me a +1 there I can merge it straight away

done

**Pull request merged**